### PR TITLE
Adding type hints for Distilbert

### PR DESCRIPTION
**Pull request by johnryan465**
# What does this PR do?

Added type hints for DistillBert

Fixes # (https://github.com/huggingface/transformers/issues/16059)

There are some slight issues re the type signatures, eg the example below where the code will fail if the head_mask paramerter is actually none, however I didn't remove the default paramerter as that would require a change in parameter order etc.

(https://github.com/johnryan465/transformers/blob/a11858c3dd7e28dec615c2d00112267ed231c657/src/transformers/models/distilbert/modeling_distilbert.py#L318) 

@Rocketknight1

### CODE REVIEW

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 1]
```diff
@@ -256,7 +267,13 @@ def __init__(self, config):
...
         self.output_layer_norm = nn.LayerNorm(normalized_shape=config.dim, eps=1e-12)
 
-    def forward(self, x, attn_mask=None, head_mask=None, output_attentions=False):
+    def forward(
+        self,
+        x: torch.Tensor,
+        attn_mask: Optional[torch.Tensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        output_attentions: bool = False,
+    ) -> Tuple[torch.Tensor, ...]:
```
Comment (Rocketknight1) [inline src/transformers/models/distilbert/modeling_distilbert.py:276]:
```suggestion
    ) -> Tuple[torch.Tensor]:
```
I think `Tuple[torch.Tensor]` covers tuples of varying lengths!

**Review by Rocketknight1 — COMMENTED** [review]
Hey, I made a bunch of small suggestions here, but it mostly looks good!

File: src/transformers/models/distilbert/modeling_distilbert.py [hunk 2]
```diff
@@ -215,7 +226,7 @@ def unshape(x):
 
         context = torch.matmul(weights, v)  # (bs, n_heads, q_length, dim_per_head)
         context = unshape(context)  # (bs, q_length, dim)
-        context = self.out_lin(context)  # (bs, q_length, dim)
+        context: torch.Tensor = self.out_lin(context)  # (bs, q_length, dim)
```
[the remark this answers is not in the document]
Comment (johnryan465) [reply src/transformers/models/distilbert/modeling_distilbert.py:229]: Added these to make MyPy happy, but will remove.

File: src/transformers/models/distilbert/modeling_distilbert.py (hunk 1 above)
Comment (johnryan465) [reply src/transformers/models/distilbert/modeling_distilbert.py:276]: Interestingly it seems MyPy doesn't complain about the difference but PyLance does!

Comment (Rocketknight1) [reply src/transformers/models/distilbert/modeling_distilbert.py:276]: Ah, okay. In that case my correction is wrong, keep the ellipsis!

**Comment by Rocketknight1** [comment]
@johnryan465 I reviewed things and you're right - `Tuple[torch.Tensor, ...]` is correct! Let me know when you're happy with the rest and I'll do another review and hopefully merge.

**Comment by johnryan465** [comment]
@Rocketknight1  I have gone through your comments and I hopefully addressed them. I'm happy enough with the code now I think. Have a good St. Patrick's Day weekend ☘️!

**Comment by Rocketknight1** [comment]
It looks perfect now, thank you! :shamrock:

**Review by Rocketknight1 — APPROVED** [review]

**Pull request merged**